circuits/plonk-plookup/src/constraints.rs

```diff
@@ -0,0 +1,386 @@
+            lkpl8,
+            lkpm,
+
+            // ECC arithmetic constraint polynomials
+            addl: addm.evaluate_over_domain_by_ref(domain.d4),
+            addm,
+            doublel: doublem.evaluate_over_domain_by_ref(domain.d8),
+            doublem,
+            mul1l: mul1m.evaluate_over_domain_by_ref(domain.d4),
+            mul2l: mul2m.evaluate_over_domain_by_ref(domain.d8),
+            mul1m,
+            mul2m,
+            emull: emulm.evaluate_over_domain_by_ref(domain.d4),
+            emulm,
+
+            // lookup table polynonial
+            table1,
+            table8w: table8.shift(8),
+            table8,
+            tablem,
+
+            // constant polynomials
+            l14: E::<F, D<F>>::from_vec_and_domain((0..domain.d4.size).map(|j| l18.evals[2*j as usize]).collect(), domain.d4),
+            l18,
+            l04: E::<F, D<F>>::from_vec_and_domain((0..domain.d4.size).map(|j| l08.evals[2*j as usize]).collect(), domain.d4),
+            l08,
+            zero4: E::<F, D<F>>::from_vec_and_domain(vec![F::zero(); domain.d4.size as usize], domain.d4),
+            zero8: E::<F, D<F>>::from_vec_and_domain(vec![F::zero(); domain.d8.size as usize], domain.d8),
+            zkpl: zkpm.evaluate_over_domain_by_ref(domain.d8),
+            zkpm,
+
+            gates,
+            shift,
+            endo: F::zero(),
+            fr_sponge_params,
+        })
+    }
+
+    // This function verifies the consistency of the wire
+    // assignements (witness) against the constraints
+    //     witness: wire assignement witness
+    //     RETURN: verification status
+    pub fn verify
+    (
+        &self,
+        witness: &[Vec<F>; COLUMNS]
+    ) -> bool
+    {
+        let p = vec![F::one(), F::zero(), F::zero(), F::zero(), F::zero(), F::zero(), F::zero()];
+        (0..self.gates.len()).all
+        (
+            |j|
+                // verify permutation consistency
+                (0..COLUMNS).all(|i|
+                {
+                    let wire = self.gates[j].wires[i];
+                    witness[i][j] == witness[wire.col][wire.row]
+                })
+                &&
+                // verify witness against constraints
+                if j < self.public {self.gates[j].c == p} else {self.gates[j].verify(witness, &self)}
+        )
+    }
+
+    // sample coordinate shifts deterministically
+    pub fn sample_shift(domain: &D<F>, i: &mut u32) -> F
+    {
+        let mut h = Blake2b::new();
+        h.input(&{*i += 1; *i}.to_be_bytes());
+        let mut r = F::from_random_bytes(&h.result()[..31]).unwrap();
+        while r.legendre().is_qnr() == false || domain.evaluate_vanishing_polynomial(r).is_zero()
+        {
+            let mut h = Blake2b::new();
+            h.input(&{*i += 1; *i}.to_be_bytes());
+            r = F::from_random_bytes(&h.result()[..31]).unwrap();
+        }
+        r
+    }
+
+    pub fn sample_shifts(domain: &D<F>, len: usize) -> Vec<F>
```

Comment:
Can we share this code with the non-plookup implementation? Maybe pull out into a utils library or something?